consensus/types/src/payload.rs

```diff
@@ -0,0 +1,236 @@
+use crate::{test_utils::TestRandom, *};
+use derivative::Derivative;
+use serde::de::DeserializeOwned;
+use serde::{Deserialize, Serialize};
+use ssz::{Decode, DecodeError, Encode};
+use std::convert::TryFrom;
+use std::fmt::Debug;
+use std::hash::Hash;
+use test_random_derive::TestRandom;
+use tree_hash::TreeHash;
+
+pub enum BlockType {
+    Blinded,
+    Full,
+}
+
+pub trait ExecPayload<T: EthSpec>:
+    Encode
+    + Decode
+    + TestRandom
+    + TreeHash
+    + Default
+    + PartialEq
+    + Serialize
+    + DeserializeOwned
+    + Hash
+    + TryFrom<ExecutionPayloadHeader<T>>
+    + From<ExecutionPayload<T>>
+{
+    fn block_type() -> BlockType;
+
+    /// Convert the payload into a payload header.
+    fn to_execution_payload_header(&self) -> ExecutionPayloadHeader<T>;
+
+    // We provide a subset of field accessors, for the fields used in `consensus`.
+    //
+    // More fields can be added here if you wish.
+    fn parent_hash(&self) -> ExecutionBlockHash;
+    fn prev_randao(&self) -> Hash256;
+    fn timestamp(&self) -> u64;
+    fn block_hash(&self) -> ExecutionBlockHash;
+}
+
+impl<T: EthSpec> ExecPayload<T> for FullPayload<T> {
+    fn block_type() -> BlockType {
+        BlockType::Full
+    }
+
+    fn to_execution_payload_header(&self) -> ExecutionPayloadHeader<T> {
+        ExecutionPayloadHeader::from(&self.execution_payload)
+    }
+
+    fn parent_hash(&self) -> ExecutionBlockHash {
+        self.execution_payload.parent_hash
+    }
+
+    fn prev_randao(&self) -> Hash256 {
+        self.execution_payload.prev_randao
+    }
+
+    fn timestamp(&self) -> u64 {
+        self.execution_payload.timestamp
+    }
+
+    fn block_hash(&self) -> ExecutionBlockHash {
+        self.execution_payload.block_hash
+    }
+}
+
+impl<T: EthSpec> ExecPayload<T> for BlindedPayload<T> {
+    fn block_type() -> BlockType {
+        BlockType::Blinded
+    }
+
+    fn to_execution_payload_header(&self) -> ExecutionPayloadHeader<T> {
+        self.execution_payload_header.clone()
+    }
+
+    fn parent_hash(&self) -> ExecutionBlockHash {
+        self.execution_payload_header.parent_hash
+    }
+
+    fn prev_randao(&self) -> Hash256 {
+        self.execution_payload_header.prev_randao
+    }
+
+    fn timestamp(&self) -> u64 {
+        self.execution_payload_header.timestamp
+    }
+
+    fn block_hash(&self) -> ExecutionBlockHash {
+        self.execution_payload_header.block_hash
+    }
+}
+
+#[derive(Default, Debug, Clone, TestRandom, Serialize, Deserialize, Derivative)]
+#[derivative(PartialEq, Hash(bound = "T: EthSpec"))]
+#[serde(transparent, bound = "T: EthSpec")]
+pub struct BlindedPayload<T: EthSpec> {
```

Comment:
One of these days we should implement `#[ssz(transparent)` and `#[tree_hash(transparent)]`

Reply:
agree 100% 😩